### DISCUSSION TITLE
SelectLargerSyntaxNode skips steps

### DISCUSSION THREAD

**Discussion by shlajin** [zed-industries/zed#19795] [Feature Requests]
Hello everyone,

One of the features I frequently use in JetBrains IDEs is "grow selection" and "shrink selection." Zed provides similar commands with `SelectLargerSyntaxNode` and `SelectSmallerSyntaxNode`, but they tend to skip certain steps, which makes them less intuitive and, overall, less consistent.

For example, consider the following TypeScript code snippet:

```
function myFunc() {
  const anotherOne = () => {
    const a: string = "hello";
    return a + ("world" + "!");
  }
}
```
If I place the caret inside world and repeatedly use SelectLargerSyntaxNode, this is the sequence of selections I get:

```
// Each step shows the text selected as I expand to a larger node
world // selected inside the quotes first
"world"  // selected outside the quotes
"world" + "!" // selected inside the parentheses
("world" + "!") // selected outside the parentheses
a + ("world" + "!")
return a + ("world" + "!");
```
So far, this behavior is perfect. However, the next selection skips directly to this:

```
{
  const a: string = "hello";
  return a + ("world" + "!");
}
```

Notice that it skips two critical selections:

1) It never selects the entire line (e.g., return a + ("world" + "!"); including the indentation). While this is not a major issue for me, it would still be useful.
2) It selects outside the { } brackets without first selecting the inner content (i.e., it jumps straight to the block without selecting everything within { } first). This inconsistency is significant.
This behavior is not consistent with how other parentheses and quotes are treated (where the content inside is selected first, followed by the outer structure).

In case you're wondering about the practical use case—when I want to modify the contents of a function or method, I usually select one of the lines within the function, expand the selection, and start typing immediately. With the current behavior, I have to delete the existing content, then manually add brackets (if I start typing a bracket without deleting first, it will just wrap the content in a new set of brackets), and then hit enter to move the caret to a new line (since the {} is now on the same line).

Is there a way to improve the SelectLargerSyntaxNode to make the selection process more consistent? Thank you.

**Comment by waitstop** [comment]
Same problem, I often used "editor.action.smartSelect" in vscode. So I'm also wondering if this can be configured somehow.

**Comment by shlajin** [comment]
I dug a little deeper and it seems like it just "walks up" the tree-sitter hierarchy, that does not include the line and the step inside the curly braces. Something tells me it dramatically complicates this request...

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
